commit/merkleroot/validate_observation.go

```diff
@@ -0,0 +1,180 @@
+package merkleroot
+
+import (
+	"context"
+	"fmt"
+
+	mapset "github.com/deckarep/golang-set/v2"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+
+	"github.com/smartcontractkit/chainlink-ccip/shared"
+
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/reader"
+	"github.com/smartcontractkit/chainlink-ccip/plugintypes"
+)
+
+func (w *Processor) ValidateObservation(
+	_ Outcome,
+	_ Query,
+	ao shared.AttributedObservation[Observation]) error {
+	obs := ao.Observation
+	if err := validateFChain(obs.FChain); err != nil {
+		return fmt.Errorf("failed to validate FChain: %w", err)
+	}
+	observerSupportedChains, err := w.chainSupport.SupportedChains(ao.OracleID)
+	if err != nil {
+		return fmt.Errorf("failed to get supported chains: %w", err)
+	}
+
+	supportsDestChain, err := w.chainSupport.SupportsDestChain(ao.OracleID)
+	if err != nil {
+		return fmt.Errorf("call to supportsDestChain failed: %w", err)
+	}
+
+	if err := validateObservedMerkleRoots(obs.MerkleRoots, ao.OracleID, observerSupportedChains); err != nil {
+		return fmt.Errorf("failed to validate MerkleRoots: %w", err)
+	}
+
+	if err := validateObservedOnRampMaxSeqNums(obs.OnRampMaxSeqNums, ao.OracleID, observerSupportedChains); err != nil {
+		return fmt.Errorf("failed to validate OnRampMaxSeqNums: %w", err)
+	}
+
+	if err := validateObservedOffRampMaxSeqNums(obs.OffRampNextSeqNums, ao.OracleID, supportsDestChain); err != nil {
+		return fmt.Errorf("failed to validate OffRampNextSeqNums: %w", err)
+	}
+
+	return nil
+}
+
+func validateObservedMerkleRoots(
+	merkleRoots []cciptypes.MerkleRootChain,
+	observer commontypes.OracleID,
+	observerSupportedChains mapset.Set[cciptypes.ChainSelector],
+) error {
+	if len(merkleRoots) == 0 {
+		return nil
+	}
+
+	seenChains := mapset.NewSet[cciptypes.ChainSelector]()
+	for _, root := range merkleRoots {
+		if !observerSupportedChains.Contains(root.ChainSel) {
+			return fmt.Errorf("found merkle root for chain %d, but this chain is not supported by Observer %d",
+				root.ChainSel, observer)
+		}
+
+		if seenChains.Contains(root.ChainSel) {
+			return fmt.Errorf("duplicate merkle root for chain %d", root.ChainSel)
```

Comment:
Ah yes, there it is. One ChainSel per slice. I think we should figure out the encoding so that we can use maps.